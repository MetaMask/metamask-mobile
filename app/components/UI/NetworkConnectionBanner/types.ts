/**
 * Network Connection Banner Status
 *
 * Slow: The network is taking more than 5 seconds to initialize.
 * Unavailable: The network is not available.
 */
export type NetworkConnectionBannerStatus = 'slow' | 'unavailable';
