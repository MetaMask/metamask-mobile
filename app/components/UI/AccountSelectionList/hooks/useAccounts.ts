// Third party dependencies.
import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { toChecksumAddress } from 'ethereumjs-util';

// External Dependencies.
import UntypedEngine from '../../../../core/Engine';
import { Account } from '..';

export const useAccounts = () => {
  const Engine = UntypedEngine as any;
  let [accounts, setAccounts] = useState<Account[]>([]);
  const identities = useSelector(
    (state: any) =>
      state.engine.backgroundState.PreferencesController.identities,
  );

  const getAccounts = () => {
    // Reading keyrings directly from Redux doesn't work at the momemt.
    const keyrings: any[] = Engine.context.KeyringController.state.keyrings;
    const latestAccounts: Account[] = keyrings.reduce((result, keyring) => {
      const { accounts: accountAddresses }: { accounts: string[] } = keyring;
      for (const address of accountAddresses) {
        const checksummedAddress = toChecksumAddress(address);
        const identity = identities[checksummedAddress];
        if (!identity) continue;
        const { name } = identity;
        const mappedAccount: Account = {
          name,
          address: checksummedAddress,
          // TODO - Also fetch assets. Reference AccountList component.
          // assets
        };
        result.push(mappedAccount);
      }
      return result;
    }, []);

    setAccounts(latestAccounts);
  };

  // Enable ability to manually fetch latest account on demand.
  const fetchLatestAccounts = (delay?: number) =>
    setTimeout(getAccounts, delay);

  useEffect(() => {
    fetchLatestAccounts();
  }, []);

  return { accounts, fetchLatestAccounts };
};
