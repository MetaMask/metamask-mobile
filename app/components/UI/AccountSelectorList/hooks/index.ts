/* eslint-disable import/prefer-default-export */

export { useAccounts } from './useAccounts';
