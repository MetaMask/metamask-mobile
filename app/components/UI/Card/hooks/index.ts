import { useCallback, useEffect } from 'react';
import { useCardSDK } from '../sdk/sdk';

/**
 * Hook to automatically check card holder status when address or card feature changes
 */
export const useCardHolderStatus = (autoFetch = false) => {
  const { isCardHolderStatus, checkCardHolderStatus, isLoadingCardHolder } =
    useCardSDK();

  useEffect(() => {
    if (autoFetch && isCardHolderStatus === null) {
      checkCardHolderStatus();
    }
  }, [autoFetch, isCardHolderStatus, checkCardHolderStatus]);

  return {
    isCardHolder: isCardHolderStatus,
    isLoading: isLoadingCardHolder,
    refetch: useCallback(
      () => checkCardHolderStatus(),
      [checkCardHolderStatus],
    ),
  };
};

/**
 * Hook to automatically fetch card token balances when address or card feature changes
 */
export const useCardTokenBalances = (autoFetch = false) => {
  const {
    tokenBalances,
    priorityToken,
    fetchBalances,
    isLoadingBalances,
    balancesError,
  } = useCardSDK();

  useEffect(() => {
    if (autoFetch) {
      fetchBalances();
    }
  }, [autoFetch, fetchBalances]);

  return {
    balances: tokenBalances,
    priorityToken,
    isLoading: isLoadingBalances,
    error: balancesError,
    refetch: useCallback(() => fetchBalances(), [fetchBalances]),
  };
};

/**
 * Hook to manage user geolocation
 */
export const useUserLocation = (autoFetch = false) => {
  const { userLocation, fetchUserLocation, isLoadingLocation } = useCardSDK();

  useEffect(() => {
    if (autoFetch && !userLocation) {
      fetchUserLocation();
    }
  }, [autoFetch, userLocation, fetchUserLocation]);

  return {
    location: userLocation,
    isLoading: isLoadingLocation,
    fetchLocation: fetchUserLocation,
  };
};
