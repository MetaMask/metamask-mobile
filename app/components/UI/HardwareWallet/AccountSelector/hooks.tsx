import { useState, useMemo, useEffect } from 'react';
import Engine from '../../../../core/Engine';

export interface IAccount {
  address: string;
}

export interface AccountBalance {
  balance: string;
}

export interface AccountBalances {
  [p: string]: AccountBalance;
}

export const useAccountsBalance = (accounts: IAccount[]) => {
  const [trackedAccounts, setTrackedAccounts] = useState<AccountBalances>({});
  const AccountTrackerController = useMemo(
    () => (Engine.context as any).AccountTrackerController,
    [],
  );

  useEffect(
    () => {
      const unTrackedAccounts: string[] = [];
      accounts.forEach((account) => {
        if (!trackedAccounts[account.address]) {
          unTrackedAccounts.push(account.address);
        }
      });
      if (unTrackedAccounts.length > 0) {
        AccountTrackerController.syncBalanceWithAddresses(
          unTrackedAccounts,
        ).then((_trackedAccounts: AccountBalances) => {
          setTrackedAccounts({
            ...trackedAccounts,
            ..._trackedAccounts,
          });
        });
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [AccountTrackerController, accounts],
  );

  return trackedAccounts;
};
