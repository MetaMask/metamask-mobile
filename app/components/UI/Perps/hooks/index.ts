// Core hooks (direct controller access)
export { usePerpsDeposit } from './usePerpsDeposit';
export { usePerpsDepositQuote } from './usePerpsDepositQuote';
export { usePerpsMarkets } from './usePerpsMarkets';
export { usePerpsNetwork } from './usePerpsNetwork';
export { usePerpsNetworkConfig } from './usePerpsNetworkConfig';
export { usePerpsTrading } from './usePerpsTrading';
export { usePerpsWithdrawQuote } from './usePerpsWithdrawQuote';

// Connection management hooks
export { usePerpsConnection } from '../providers/PerpsConnectionProvider';

// State hooks (Redux selectors)
export { usePerpsAccount } from './usePerpsAccount';

// Live data hooks (WebSocket subscriptions)
export { usePerpsPositionData } from './usePerpsPositionData';
export { usePerpsPrices } from './usePerpsPrices';

// Asset metadata hooks
export { usePerpsAssetMetadata } from './usePerpsAssetsMetadata';

// Payment token hooks
export { usePerpsPaymentTokens } from './usePerpsPaymentTokens';

// Market data and calculation hooks
export { usePerpsLiquidationPrice } from './usePerpsLiquidationPrice';
export { usePerpsMarketData } from './usePerpsMarketData';
export { usePerpsMarketStats } from './usePerpsMarketStats';

// Withdrawal specific hooks
export { useWithdrawTokens } from './useWithdrawTokens';
export { useWithdrawValidation } from './useWithdrawValidation';

// UI utility hooks
export { useBalanceComparison } from './useBalanceComparison';
export { useColorPulseAnimation } from './useColorPulseAnimation';
export { usePerpsPositions } from './usePerpsPositions';
export { usePerpsTPSLUpdate } from './usePerpsTPSLUpdate';
export { usePerpsClosePosition } from './usePerpsClosePosition';
export { usePerpsOrderFees, formatFeeRate } from './usePerpsOrderFees';
export { useHasExistingPosition } from './useHasExistingPosition';
export { useMinimumOrderAmount } from './useMinimumOrderAmount';
export { usePerpsOrderForm } from './usePerpsOrderForm';
export { usePerpsOrderValidation } from './usePerpsOrderValidation';
export { usePerpsClosePositionValidation } from './usePerpsClosePositionValidation';
export { usePerpsOrderExecution } from './usePerpsOrderExecution';

// Transaction data hooks
export { usePerpsOrderFills } from './usePerpsOrderFills';
export { usePerpsOrders } from './usePerpsOrders';
export { usePerpsFunding } from './usePerpsFunding';

// Block explorer hook
export { usePerpsBlockExplorerUrl } from './usePerpsBlockExplorerUrl';

// - usePerpsDeposit (deposit flows)
// - usePerpsPendingOrders (order management)
// - usePerpsError (error handling)
// - usePerpsLivePositions (live position tracking)
// - usePerpsPositionPnL (P&L calculations)
// - usePerpsReadiness (readiness checks)
// - usePerpsOrderManagement (order workflows)
// - useStableArray (utility hook for stable array references)
