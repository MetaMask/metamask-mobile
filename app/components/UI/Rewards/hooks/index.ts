export { useSeasonStatus } from './useSeasonStatus';
export type {
  UseSeasonStatusResult,
  UseSeasonStatusOptions,
} from './useSeasonStatus';

export { useRewardsSubscription } from './useRewardsSubscription';
export type { UseRewardsSubscriptionResult } from './useRewardsSubscription';

export { useCurrentSeasonStatus } from './useCurrentSeasonStatus';
export type {
  UseCurrentSeasonStatusResult,
  UseCurrentSeasonStatusOptions,
} from './useCurrentSeasonStatus';

export { useRewardsSyncWithEngine } from './useRewardsSyncWithEngine';

export { useRewardsStore } from './useRewardsStore';
export type { UseRewardsStoreResult } from './useRewardsStore';

export { useReferralDetails } from './useReferralDetails';
export type { UseReferralDetailsResult } from './useReferralDetails';

export { useRewardsGeoLocation } from './useRewardsGeoLocation';
export type { UseRewardsGeoLocationResult } from './useRewardsGeoLocation';

export { useValidateReferralCode } from './useValidateReferralCode';
export type { UseValidateReferralCodeResult } from './useValidateReferralCode';
