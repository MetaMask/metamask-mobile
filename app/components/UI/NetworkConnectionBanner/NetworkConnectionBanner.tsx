import React from 'react';
import { View } from 'react-native';
import { useTailwind } from '@metamask/design-system-twrnc-preset';
import Banner, {
  BannerAlertSeverity,
  BannerVariant,
} from '../../../component-library/components/Banners/Banner';
import Icon, {
  IconName,
  IconSize,
} from '../../../component-library/components/Icons/Icon';
import { ButtonVariants } from '../../../component-library/components/Buttons/Button';
import Spinner, { SpinnerSize } from '../AnimatedSpinner';
import { useNetworkConnectionBanners } from '../../hooks/useNetworkConnectionBanners';
import { strings } from '../../../../locales/i18n';

/**
 * Network Connection Banner
 *
 * Shows when any network takes more than 5 seconds to initialize or is not available.
 */
const NetworkConnectionBanner: React.FC = () => {
  const { networkConnectionBannersState, currentNetwork, updateRpc } =
    useNetworkConnectionBanners();
  const tw = useTailwind();

  if (!networkConnectionBannersState.visible || !currentNetwork) {
    return null;
  }

  return (
    <View style={tw.style('px-4', 'mt-4')}>
      <Banner
        variant={BannerVariant.Alert}
        severity={BannerAlertSeverity.Warning}
        startAccessory={
          networkConnectionBannersState.status === 'slow' ? (
            <Spinner size={SpinnerSize.SM} />
          ) : (
            <Icon name={IconName.Danger} size={IconSize.Md} />
          )
        }
        title={strings(
          networkConnectionBannersState.status === 'slow'
            ? 'network_connection_banner.still_connecting_network'
            : 'network_connection_banner.unable_to_connect_network',
          {
            networkName: currentNetwork.name,
          },
        )}
        actionButtonProps={{
          variant: ButtonVariants.Link,
          label: strings('network_connection_banner.update_rpc'),
          onPress: updateRpc,
        }}
      />
    </View>
  );
};

export default NetworkConnectionBanner;
