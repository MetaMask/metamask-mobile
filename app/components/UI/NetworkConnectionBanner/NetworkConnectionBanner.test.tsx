import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import NetworkConnectionBanner from './NetworkConnectionBanner';
import { useNetworkConnectionBanner } from '../../hooks/useNetworkConnectionBanner';

jest.mock('../../hooks/useNetworkConnectionBanner');

jest.mock('../AnimatedSpinner', () => ({
  __esModule: true,
  default: ({ size }: { size: Record<string, string> }) => {
    const { View, Text } = jest.requireActual('react-native');
    return (
      <View testID="animated-spinner">
        <Text testID="spinner-size">{size}</Text>
      </View>
    );
  },
  SpinnerSize: {
    SM: 'SM',
  },
}));

jest.mock('../../../component-library/components/Icons/Icon', () => ({
  __esModule: true,
  default: ({ size }: { size: Record<string, string> }) => {
    const { View, Text } = jest.requireActual('react-native');
    return (
      <View testID="icon">
        <Text testID="icon-size">{size}</Text>
      </View>
    );
  },
  IconName: {
    Danger: 'Danger',
  },
  IconSize: {
    Md: 'Md',
  },
  IconColor: {
    Default: 'Default',
  },
}));

const mockuseNetworkConnectionBanner =
  useNetworkConnectionBanner as jest.MockedFunction<
    typeof useNetworkConnectionBanner
  >;

describe('NetworkConnectionBanner', () => {
  const mockUpdateRpc = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('snapshots', () => {
    it('should match snapshot when banner is not visible', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: { visible: false },
        updateRpc: mockUpdateRpc,
      });

      const { toJSON } = render(<NetworkConnectionBanner />);

      expect(toJSON()).toMatchSnapshot();
    });

    const statusSnapshotTestCases = [
      {
        status: 'slow' as const,
        name: 'for slow status banner',
      },
      {
        status: 'unavailable' as const,
        name: 'for unavailable status banner',
      },
    ];

    it.each(statusSnapshotTestCases)(
      'should match snapshot $name',
      ({ status }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { toJSON } = render(<NetworkConnectionBanner />);

        expect(toJSON()).toMatchSnapshot();
      },
    );

    it('should match snapshot for different network', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x89',
          status: 'slow',
          networkName: 'Polygon Mainnet',
          rpcUrl: 'https://polygon-rpc.com',
        },
        updateRpc: mockUpdateRpc,
      });

      const { toJSON } = render(<NetworkConnectionBanner />);

      expect(toJSON()).toMatchSnapshot();
    });
  });

  describe('when banner is not visible', () => {
    it('should not render when visible is false', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: { visible: false },
        updateRpc: mockUpdateRpc,
      });

      const { queryByTestId } = render(<NetworkConnectionBanner />);

      expect(queryByTestId('animated-spinner')).toBeNull();
    });
  });

  describe('when banner is visible', () => {
    const statusTestCases = [
      {
        status: 'slow' as const,
        expectedMessage: 'Still connecting to Ethereum Mainnet...',
        expectedIconTestId: 'animated-spinner',
      },
      {
        status: 'unavailable' as const,
        expectedMessage: 'Unable to connect to Ethereum Mainnet.',
        expectedIconTestId: 'icon',
      },
    ];

    it.each(statusTestCases)(
      'should render the banner with correct structure for $status status',
      ({ status, expectedMessage, expectedIconTestId }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByTestId, getByText } = render(<NetworkConnectionBanner />);

        expect(getByTestId(expectedIconTestId)).toBeTruthy();
        expect(getByText(expectedMessage)).toBeTruthy();
        expect(getByText('Update RPC')).toBeTruthy();
      },
    );

    it.each(statusTestCases)(
      'should display network name in the message for $status status',
      ({ status, expectedMessage }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByText } = render(<NetworkConnectionBanner />);

        expect(getByText(expectedMessage)).toBeTruthy();
      },
    );

    it.each(statusTestCases)(
      'should call updateRpc when Update RPC button is pressed for $status status',
      ({ status }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByText } = render(<NetworkConnectionBanner />);

        const updateButton = getByText('Update RPC');
        fireEvent.press(updateButton);

        expect(mockUpdateRpc).toHaveBeenCalledWith(
          'https://mainnet.infura.io/v3/test',
          status,
          '0x1',
        );
      },
    );

    it.each(statusTestCases)(
      'should render button with correct variant and properties for $status status',
      ({ status }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByText } = render(<NetworkConnectionBanner />);

        const updateButton = getByText('Update RPC');
        expect(updateButton).toBeTruthy();
      },
    );

    describe('status transitions', () => {
      it('should update message when status changes from slow to unavailable', () => {
        // Start with slow status
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status: 'slow',
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { rerender, getByText } = render(<NetworkConnectionBanner />);

        expect(
          getByText('Still connecting to Ethereum Mainnet...'),
        ).toBeTruthy();

        // Change to unavailable status
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status: 'unavailable',
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        rerender(<NetworkConnectionBanner />);

        expect(
          getByText('Unable to connect to Ethereum Mainnet.'),
        ).toBeTruthy();
      });

      it('should update message when status changes from unavailable to slow', () => {
        // Start with unavailable status
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status: 'unavailable',
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { rerender, getByText } = render(<NetworkConnectionBanner />);

        expect(
          getByText('Unable to connect to Ethereum Mainnet.'),
        ).toBeTruthy();

        // Change to slow status
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status: 'slow',
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        rerender(<NetworkConnectionBanner />);

        expect(
          getByText('Still connecting to Ethereum Mainnet...'),
        ).toBeTruthy();
      });
    });
  });

  describe('with different network configurations', () => {
    it('should display different network names correctly', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x89',
          status: 'slow',
          networkName: 'Polygon Mainnet',
          rpcUrl: 'https://polygon-rpc.com',
        },
        updateRpc: mockUpdateRpc,
      });

      const { getByText } = render(<NetworkConnectionBanner />);

      expect(getByText('Still connecting to Polygon Mainnet...')).toBeTruthy();
    });

    it('should handle network with special characters in name', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x1',
          status: 'slow',
          networkName: 'Test-Network (Beta)',
          rpcUrl: 'https://mainnet.infura.io/v3/test',
        },
        updateRpc: mockUpdateRpc,
      });

      const { getByText } = render(<NetworkConnectionBanner />);

      expect(
        getByText('Still connecting to Test-Network (Beta)...'),
      ).toBeTruthy();
    });

    it('should handle network with very long name', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x1',
          status: 'slow',
          networkName: 'Very Long Network Name That Might Cause Layout Issues',
          rpcUrl: 'https://mainnet.infura.io/v3/test',
        },
        updateRpc: mockUpdateRpc,
      });

      const { getByText } = render(<NetworkConnectionBanner />);

      expect(
        getByText(
          'Still connecting to Very Long Network Name That Might Cause Layout Issues...',
        ),
      ).toBeTruthy();
    });
  });

  describe('accessibility', () => {
    const accessibilityTestCases = [
      {
        status: 'slow' as const,
        expectedMessage: 'Still connecting to Ethereum Mainnet...',
      },
      {
        status: 'unavailable' as const,
        expectedMessage: 'Unable to connect to Ethereum Mainnet.',
      },
    ];

    it.each(accessibilityTestCases)(
      'should render with proper accessibility structure for $status status',
      ({ status, expectedMessage }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByText } = render(<NetworkConnectionBanner />);

        // The banner should be accessible with proper text content
        expect(getByText(expectedMessage)).toBeTruthy();
        expect(getByText('Update RPC')).toBeTruthy();
      },
    );

    it.each(accessibilityTestCases)(
      'should have accessible button for updating RPC for $status status',
      ({ status }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: 'Ethereum Mainnet',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByText } = render(<NetworkConnectionBanner />);

        const updateButton = getByText('Update RPC');
        expect(updateButton).toBeTruthy();

        // Test that button is pressable
        fireEvent.press(updateButton);
        expect(mockUpdateRpc).toHaveBeenCalledWith(
          'https://mainnet.infura.io/v3/test',
          status,
          '0x1',
        );
      },
    );
  });

  describe('edge cases', () => {
    const emptyNameTestCases = [
      {
        status: 'slow' as const,
        expectedMessage: 'Still connecting to ...',
      },
      {
        status: 'unavailable' as const,
        expectedMessage: 'Unable to connect to .',
      },
    ];

    it.each(emptyNameTestCases)(
      'should handle network with empty name for $status status',
      ({ status, expectedMessage }) => {
        mockuseNetworkConnectionBanner.mockReturnValue({
          networkConnectionBannerState: {
            visible: true,
            chainId: '0x1',
            status,
            networkName: '',
            rpcUrl: 'https://mainnet.infura.io/v3/test',
          },
          updateRpc: mockUpdateRpc,
        });

        const { getByText } = render(<NetworkConnectionBanner />);

        expect(getByText(expectedMessage)).toBeTruthy();
      },
    );

    it('should handle multiple rapid button presses', () => {
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x1',
          status: 'slow',
          networkName: 'Ethereum Mainnet',
          rpcUrl: 'https://mainnet.infura.io/v3/test',
        },
        updateRpc: mockUpdateRpc,
      });

      const { getByText } = render(<NetworkConnectionBanner />);

      const updateButton = getByText('Update RPC');

      // Press button multiple times rapidly
      fireEvent.press(updateButton);
      fireEvent.press(updateButton);
      fireEvent.press(updateButton);

      expect(mockUpdateRpc).toHaveBeenCalledTimes(3);
      expect(mockUpdateRpc).toHaveBeenCalledWith(
        'https://mainnet.infura.io/v3/test',
        'slow',
        '0x1',
      );
    });
  });

  describe('component lifecycle', () => {
    it('should update when hook values change', () => {
      // Initially not visible
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: { visible: false },
        updateRpc: mockUpdateRpc,
      });

      const { rerender, queryByTestId, getByTestId } = render(
        <NetworkConnectionBanner />,
      );

      expect(queryByTestId('animated-spinner')).toBeNull();

      // Make it visible
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x1',
          status: 'slow',
          networkName: 'Ethereum Mainnet',
          rpcUrl: 'https://mainnet.infura.io/v3/test',
        },
        updateRpc: mockUpdateRpc,
      });

      rerender(<NetworkConnectionBanner />);

      expect(getByTestId('animated-spinner')).toBeTruthy();
    });

    it('should handle hook returning different network', () => {
      // Initially Ethereum
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x1',
          status: 'slow',
          networkName: 'Ethereum Mainnet',
          rpcUrl: 'https://mainnet.infura.io/v3/test',
        },
        updateRpc: mockUpdateRpc,
      });

      const { rerender, getByText } = render(<NetworkConnectionBanner />);

      expect(getByText('Still connecting to Ethereum Mainnet...')).toBeTruthy();

      // Switch to Polygon
      mockuseNetworkConnectionBanner.mockReturnValue({
        networkConnectionBannerState: {
          visible: true,
          chainId: '0x89',
          status: 'slow',
          networkName: 'Polygon Mainnet',
          rpcUrl: 'https://polygon-rpc.com',
        },
        updateRpc: mockUpdateRpc,
      });

      rerender(<NetworkConnectionBanner />);

      expect(getByText('Still connecting to Polygon Mainnet...')).toBeTruthy();
    });
  });
});
