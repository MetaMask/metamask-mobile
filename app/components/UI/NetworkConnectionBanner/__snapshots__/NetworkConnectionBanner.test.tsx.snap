// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkConnectionBanner snapshots should match snapshot for different network 1`] = `
<View
  style={
    {
      "marginTop": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#9a63001a",
        "borderColor": "#9a6300",
        "borderLeftWidth": 4,
        "borderRadius": 4,
        "flexDirection": "row",
        "padding": 12,
        "paddingLeft": 8,
      }
    }
    testID="banneralert"
    variant="Alert"
  >
    <View
      style={
        {
          "marginRight": 8,
        }
      }
    >
      <View
        testID="animated-spinner"
      >
        <Text
          testID="spinner-size"
        >
          SM
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#121314",
            "fontFamily": "Geist Medium",
            "fontSize": 18,
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        Still connecting to Polygon Mainnet...
      </Text>
      <Text
        accessibilityRole="link"
        accessible={true}
        onPress={[MockFunction]}
        onPressIn={[Function]}
        onPressOut={[Function]}
        style={
          {
            "backgroundColor": "transparent",
            "color": "#121314",
            "fontFamily": "Geist Regular",
            "fontSize": 16,
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
        suppressHighlighting={true}
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#4459ff",
              "fontFamily": "Geist Medium",
              "fontSize": 16,
              "letterSpacing": 0,
              "lineHeight": 24,
            }
          }
        >
          Update RPC
        </Text>
      </Text>
    </View>
  </View>
</View>
`;

exports[`NetworkConnectionBanner snapshots should match snapshot for slow status banner 1`] = `
<View
  style={
    {
      "marginTop": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#9a63001a",
        "borderColor": "#9a6300",
        "borderLeftWidth": 4,
        "borderRadius": 4,
        "flexDirection": "row",
        "padding": 12,
        "paddingLeft": 8,
      }
    }
    testID="banneralert"
    variant="Alert"
  >
    <View
      style={
        {
          "marginRight": 8,
        }
      }
    >
      <View
        testID="animated-spinner"
      >
        <Text
          testID="spinner-size"
        >
          SM
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#121314",
            "fontFamily": "Geist Medium",
            "fontSize": 18,
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        Still connecting to Ethereum Mainnet...
      </Text>
      <Text
        accessibilityRole="link"
        accessible={true}
        onPress={[MockFunction]}
        onPressIn={[Function]}
        onPressOut={[Function]}
        style={
          {
            "backgroundColor": "transparent",
            "color": "#121314",
            "fontFamily": "Geist Regular",
            "fontSize": 16,
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
        suppressHighlighting={true}
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#4459ff",
              "fontFamily": "Geist Medium",
              "fontSize": 16,
              "letterSpacing": 0,
              "lineHeight": 24,
            }
          }
        >
          Update RPC
        </Text>
      </Text>
    </View>
  </View>
</View>
`;

exports[`NetworkConnectionBanner snapshots should match snapshot for unavailable status banner 1`] = `
<View
  style={
    {
      "marginTop": 16,
      "paddingLeft": 16,
      "paddingRight": 16,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#9a63001a",
        "borderColor": "#9a6300",
        "borderLeftWidth": 4,
        "borderRadius": 4,
        "flexDirection": "row",
        "padding": 12,
        "paddingLeft": 8,
      }
    }
    testID="banneralert"
    variant="Alert"
  >
    <View
      style={
        {
          "marginRight": 8,
        }
      }
    >
      <View
        testID="icon"
      >
        <Text
          testID="icon-size"
        >
          Md
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#121314",
            "fontFamily": "Geist Medium",
            "fontSize": 18,
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
      >
        Unable to connect to Ethereum Mainnet.
      </Text>
      <Text
        accessibilityRole="link"
        accessible={true}
        onPress={[MockFunction]}
        onPressIn={[Function]}
        onPressOut={[Function]}
        style={
          {
            "backgroundColor": "transparent",
            "color": "#121314",
            "fontFamily": "Geist Regular",
            "fontSize": 16,
            "letterSpacing": 0,
            "lineHeight": 24,
          }
        }
        suppressHighlighting={true}
      >
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#4459ff",
              "fontFamily": "Geist Medium",
              "fontSize": 16,
              "letterSpacing": 0,
              "lineHeight": 24,
            }
          }
        >
          Update RPC
        </Text>
      </Text>
    </View>
  </View>
</View>
`;

exports[`NetworkConnectionBanner snapshots should match snapshot when banner is not visible 1`] = `null`;
